Publish invalid messages to a validation errors topic

zoidyzoidzoid opened the pull request:
Adds `validation_error_callback` to `publish`:

e.g.:

```
        def validation_error_callback(event, exception, protocol):
            message = {
                'meta': {
                    'date': datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%SZ'),
                    'hostname': gethostname(),
                    'service': 'example-app',
                    'uuid': str(uuid4())
                },
                'schema': schema,
                'errors': [err.message for err in exception.errors]
            }
            protocol.publish(topic, message)
```

zoidyzoidzoid committed:
Publish validation error events

zoidyzoidzoid committed:
Correct errors schema attribute

zoidyzoidzoid committed:
Use a validation error handler instead

zoidyzoidzoid committed:
Test for invalid validation error messages

zoidyzoidzoid committed:
I don't like this / need to refactor it

bertuss commented on tests/test_protocols.py:
```diff
@@ -106,3 +113,83 @@ def test_invalid_message(self):
+    @classmethod
+    def setUp(cls):
+        cls.schema = {
+            'required': ['foo']
```
Shouldn't this be `'required': ['familyName']`?

zoidyzoidzoid replied:
Yeah, this `cls.schema` thing doesn't work anymore. Just using that public example one now.

zoidyzoidzoid replied:
Thanks though, will clean this up.

zoidyzoidzoid committed:
We don't allow overriding schema anymore

It didn't help do what I wanted, where we could override one schema but not the validation error schema

bertuss commented on tests/test_protocols.py:
```diff
@@ -106,3 +113,79 @@ def test_invalid_message(self):
+
+    def test_invalid_message(self):
+        topic = 'validation_error'
+        schema = 'http://schema.superbalist.com/events/validation_error/validation_error/1.0.json'
```
I know these are public but I don't think we should use our company schema to test.
Maybe we could serve the `validation_error` schema file with gh-pages on this repo?

zoidyzoidzoid replied:
That's a good idea. I'm not a fan either.

bertuss commented on pubsub/protocol.py:
```diff
@@ -1,20 +1,34 @@
         self.filter = filter
 
-    def publish(self, topic, message):
+    def publish(self, topic, message, validation_error_message=False):
```
Bit nitpicky, but I would call `validation_error_message` something else like `suppress_error_handler`.

zoidyzoidzoid replied:
Yeah, then it becomes more general though, which isn't the worst.

bertuss reviewed (approved):
A couple of comments, but otherwise it looks good to me.

zoidyzoidzoid commented on pubsub/validators/validator.py:
```diff
@@ -16,18 +29,21 @@ def validate_message(self, message):
-            jsonschema.validate(message, schema)
-        else:
-            raise ValidationError('Incorrect schema uri')
+        matches = re.findall(r'(.+)://(.+/)?(.+)\.json', schema_uri)
```
There is probably a simpler way to assert URL using [`urlparse`](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlparse) or something.

zoidyzoidzoid replied:
I looked into how Django validates URLs and it uses it's own vendored regex :sob:

zoidyzoidzoid committed:
Update test_protocols.py

zoidyzoidzoid committed:
Make the validation error callback more like other callbacks
